scalding-core/src/main/scala/com/twitter/scalding/estimation/memory/MemoryHistoryEstimator.scala

```diff
@@ -0,0 +1,99 @@
+  import MemoryHistoryEstimator.MemoryRichTask
+
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
+  override def maxHistory(conf: JobConf): Int = MemoryEstimatorConfig.getMaxHistory(conf)
+
+  override protected def estimate(info: FlowStrategyInfo, conf: JobConf, history: Seq[FlowStepHistory]): Option[MemoryEstimate] = {
+    // iterate over mem history
+    // collect: for maps, list of max memory in past runs
+    //          for reduce, list of max memory in past runs
+    // compute smoothed memory est
+    // multiple by scale factor
+    // cap estimate to max size if needed
+    // handle gc
+    // return
+    val alpha = MemoryEstimatorConfig.getAlpha(conf)
+    val scaleFactor = MemoryEstimatorConfig.getScaleFactor(conf)
+    val yarnSchedulerIncrementMB = MemoryEstimatorConfig.getYarnSchedulerIncrement(conf)
+
+    LOG.info(s"Attempting to estimate memory settings with alpha: $alpha, scaleFactor: $scaleFactor, schedulerIncrement: $yarnSchedulerIncrementMB")
+
+    val maxMemoryList: Seq[(Option[Long], Option[Long])] = history.map{ h => getMapReduceMemory(h) }
+    val maxMapList: Seq[Long] = maxMemoryList.flatMap(_._1)
+    val maxReduceList: Seq[Long] = maxMemoryList.flatMap(_._2)
+
+    val mapSmoothEst: Double = smoothedAverage(maxMapList, alpha)
+    val reduceSmoothEst: Double = smoothedAverage(maxReduceList, alpha)
+
+    val mapScaledEst = mapSmoothEst * scaleFactor
+    val reduceScaledEst = reduceSmoothEst * scaleFactor
+
+    //todo handle gc values
+    val cappedMapEst = cappedMemoryEstimateMB(mapScaledEst)
+    val cappedReduceEst = cappedMemoryEstimateMB(reduceScaledEst)
+
+    LOG.info(s"Calculated map val for: $maxMapList smoothAvg = $mapSmoothEst, scaled: $mapScaledEst, capped: $cappedMapEst")
+    LOG.info(s"Calculated reduce val for: $maxReduceList smoothAvg = $reduceSmoothEst, scaled: $reduceScaledEst, capped: $cappedReduceEst")
+
+    Some(MemoryEstimate(cappedMapEst, cappedReduceEst))
+  }
+
+  private def getMapReduceMemory(history: FlowStepHistory): (Option[Long], Option[Long]) = {
+    LOG.info(s"Processing tasks: ${history.tasks}")
+    val reduceTasks: Seq[Task] = history.tasks.filter { t => t.taskType.contains("REDUCE") }
+    val mapTasks: Seq[Task] = history.tasks.filter { t => t.taskType.contains("MAP") }
+
+    // handle empty task list due to either no task history / lack of reducers
+    val maxReduceCommittedHeap: Option[Long] = if (reduceTasks.isEmpty) None else Some(reduceTasks.flatMap(_.committedHeapBytes).max)
+    val maxMapCommittedHeap: Option[Long] = if (mapTasks.isEmpty) None else Some(mapTasks.flatMap(_.committedHeapBytes).max)
+
+    LOG.info(s"Calculated max committed heap for job: ${history.keys}, map: $maxMapCommittedHeap reduce: $maxReduceCommittedHeap")
+    (maxMapCommittedHeap, maxReduceCommittedHeap)
+  }
+
+  // memoryEstimate = (currentMemoryValue * alpha) + (1 - alpha) * oldEstimate
+  private def smoothedAverage(memoryList: Seq[Long], alpha: Double): Double = {
```

Comment:
not sure, most of the private methods can be static, but I don't see why we need to move them.